test/hotspot/jtreg/serviceability/jvmti/vthread/GetThreadStateMountedTest/libGetThreadStateMountedTest.cpp

```diff
@@ -0,0 +1,212 @@
+ * questions.
+ */
+
+#include <jni.h>
+#include <jvmti.h>
+#include <jvmti_common.h>
+#include <atomic>
+
+static jvmtiEnv *jvmti = nullptr;
+static jint error_count = 0;
+
+extern "C" JNIEXPORT jint JNICALL
+Agent_OnLoad(JavaVM *vm, char *options, void *reserved) {
+  if (vm->GetEnv((void **)&jvmti, JVMTI_VERSION) != JNI_OK) {
+    LOG("Could not initialize JVMTI\n");
+    return JNI_ERR;
+  }
+  jvmtiCapabilities caps;
+  memset(&caps, 0, sizeof(caps));
+  caps.can_support_virtual_threads = 1;
+  caps.can_suspend = 1;
+  caps.can_signal_thread = 1;
+  jvmtiError err = jvmti->AddCapabilities(&caps);
+  if (err != JVMTI_ERROR_NONE) {
+    LOG("JVMTI AddCapabilities error: %d\n", err);
+    return JNI_ERR;
+  }
+
+  return JNI_OK;
+}
+
+
+extern "C" JNIEXPORT jboolean JNICALL
+Java_GetThreadStateMountedTest_trySuspendInWaitingState(JNIEnv* jni, jclass clazz, jthread vthread) {
+  const int max_retries = 10;
+  for (int i = 0; i < max_retries; i++) {
+    // wait a bit
+    sleep_ms(100);
+
+    // suspend the thread
+    LOG("suspend vthread (%d)\n", i);
+    suspend_thread(jvmti, jni, vthread);
+
+    jint state = get_thread_state(jvmti, jni, vthread);
+    if ((state & JVMTI_THREAD_STATE_WAITING) != 0) {
+      LOG("suspended in WAITING state\n");
+      return JNI_TRUE;
+    }
+    LOG("suspended vthread is not waiting: state = %x (%s)\n", state, TranslateState(state));
+    LOG("resume vthread\n");
+    resume_thread(jvmti, jni, vthread);
+  }
+  LOG("ERROR: failed to suspend in WAITING state in %d tries\n", max_retries);
+  return JNI_FALSE;
+
+}
+
+static void verify_thread_state(const char *name, JNIEnv* jni,
+  jthread thread, jint expected_strong, jint expected_weak)
+{
+  jint state = get_thread_state(jvmti, jni, thread);
+  LOG("%s state(%x): %s\n", name, state, TranslateState(state));
+  bool failed = false;
+  // check 1: all expected_strong bits are set
+  jint actual_strong = state & expected_strong;
+  if (actual_strong != expected_strong) {
+    failed = true;
+    jint missed = expected_strong - actual_strong;
+    LOG("  ERROR: some mandatory bits are not set (%x): %s\n",
+        missed, TranslateState(missed));
+  }
+  // check 2: no bits other than (expected_strong | expected_weak) are set
+  jint actual_full = state & (expected_strong | expected_weak);
+  if (actual_full != state) {
+    failed = true;
+    jint unexpected = state - actual_full;
+    LOG("  ERROR: some unexpected bits are set (%x): %s\n",
+        unexpected, TranslateState(unexpected));
+  }
+  // check 3: expected_weak checks
+  if (expected_weak != 0) {
+    // check 3a: at least 1 bit from expected_weak is set
+    if ((state & expected_weak) == 0) {
+      failed = true;
+      LOG("  ERROR: no expected 'weak' bits are set\n");
+    }
+    // check 3b: not all expected_weak bits are set
+    if ((state & expected_weak) == expected_weak) {
+      failed = true;
+      LOG("  ERROR: all expected 'weak' bits are set\n");
+    }
+  }
```

Comment:
Nit: I guess, error message at line 110 has to be: 
`ERROR: not all expected 'weak' bits are set`
Also, it looks like the check 3a is not really needed as the 3b should cover it.
But I leave it up to you, if you think check 3a gives some convenience.

Reply:
'weak' value is for 'sleeping' testcase. In the case we have 2 valid states: JVMTI_THREAD_STATE_SLEEPING and JVMTI_THREAD_STATE_PARKED, but only one of them should be reported.
Test passes expected_weak == JVMTI_THREAD_STATE_SLEEPING | JVMTI_THREAD_STATE_PARKED.
check 3a verifies that at least 1 bit is set (JVMTI_THREAD_STATE_SLEEPING, JVMTI_THREAD_STATE_PARKED or both).
check 3b verifies that only one of the bits is set (i.e. we don't have both JVMTI_THREAD_STATE_SLEEPING and JVMTI_THREAD_STATE_PARKED)